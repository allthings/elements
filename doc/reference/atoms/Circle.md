<!-- 
This is an auto-generated markdown. 
You can change it in "src/atoms/Circle.tsx" and run build:docs to update this file.
-->
# Circle
Simple circle element that is used to build other things.

Check out ProfileImage to see it in action.

```example
<ThemeProvider>
   <ResourceProvider>
     <View direction="row" alignV="center">
       <Circle outline fill={false} outlineColor="lightGrey">
         <Icon size="s" name="remove-filled" color="lightGrey" />
       </Circle>
       <Circle color="primary" />
       <Circle color="#bada55" radius={20} />
       <Circle color="gray" radius={10} />
     </View>
   </ResourceProvider>
</ThemeProvider>
```
## Usage
| Name        | Type           | Description  |
| ----------- |:--------------:| ------------:|
|children|ReactNode|Content of the Circle will be always centered
|color|string|Color of the Circle. Allows theme names (like 'primary') or hex colors
|radius|number|Radius of Circle, defaults to 40
|outline|boolean|If true it will only render the outline
|outlineColor|string|If passed outline will use as outline color instead of color
|fill|boolean|If true it will fill the circle
Passing true, will make the view fill out available space
