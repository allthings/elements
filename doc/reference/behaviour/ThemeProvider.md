<!-- 
This is an auto-generated markdown. 
You can change it in "src/behaviour/ThemeProvider.jsx" and run build:docs to update this file.
-->
# ThemeProvider
All elements support theming by default, and therefore every element must be wrapped inside a ThemeProvider.
The ThemeProvider allows you to define the default colors for most elements.

**Example**: If you want all you buttons to be red, instead of writing <Button color="red" /> all the time, you might want to set the "primary" color of your theme to red.
## Usage
| Name        | Type           | Description  |
| ----------- |:--------------:| ------------:|
|root|bool|
|theme|shape|<br>Default: {
  primary: ColorPalette.primary.blue,
  text: ColorPalette.text.primary,
  secondaryText: ColorPalette.text.secondary,
  titleColor: ColorPalette.text.primary,
  contrast: ColorPalette.white,
  warn: ColorPalette.red,
  disabled: ColorPalette.grey,
  background: ColorPalette.background.bright,
  textOnBackground: ColorPalette.white,
}
|children|node|
