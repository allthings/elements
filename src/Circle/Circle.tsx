import React, { FunctionComponent } from 'react'
import View, { IViewProps } from '../View'
import { css } from 'glamor'
import Theme from '../Theme'

const circle = (
  backgroundColor: string,
  outline: boolean,
  outlineColor: string,
  fill: boolean,
  radius: number,
) =>
  css({
    border: outline && `1px solid ${outlineColor || backgroundColor}`,
    backgroundColor: fill && backgroundColor,
    borderRadius: '50%',
    height: radius,
    width: radius,
    transition: 'background-color 120ms linear',
  })

interface ICircleProps extends IViewProps {
  /** Content of the Circle will be always centered */
  children?: React.ReactNode
  /** Color of the Circle. Allows theme names (like 'primary') or hex colors  */
  color?: string
  /** Radius of Circle, defaults to 40 */
  radius?: number
  /** If true it will only render the outline */
  outline?: boolean
  /** If passed outline will use as outline color instead of color */
  outlineColor?: string
  /** If true it will fill the circle */
  fill?: boolean
}

/**
 * Simple circle element that is used to build other things...
 *
 * Check out ProfileImage to see it in action.
 *
 * ```example
 * <ThemeProvider>
 *   <ResourceProvider>
 *     <View direction="row" alignV="center">
 *       <Circle outline fill={false} outlineColor="lightGrey">
 *         <Icon size="s" name="remove-filled" color="lightGrey" />
 *       </Circle>
 *       <Circle color="primary" />
 *       <Circle color="#bada55" radius={20} />
 *       <Circle color="gray" radius={10} />
 *     </View>
 *   </ResourceProvider>
 * </ThemeProvider>
 * ```
 **/
const Circle: FunctionComponent<ICircleProps> = ({
  fill = true,
  color = 'primary',
  outline = false,
  outlineColor = 'primary',
  radius = 40,
  children,
  ...props
}) => (
  <Theme>
    {({ colorize }) => (
      <View
        {...circle(
          colorize(color),
          outline,
          colorize(outlineColor),
          fill,
          radius,
        )}
        alignV="center"
        alignH="center"
        direction="column"
        {...props}
      >
        {children}
      </View>
    )}
  </Theme>
)

export default Circle
