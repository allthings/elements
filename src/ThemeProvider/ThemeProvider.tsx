import React, { FunctionComponent, createContext } from 'react'
import { ColorPalette } from '@allthings/colors'

export const defaultTheme: ITheme = {
  primary: ColorPalette.primary.blue,
  text: ColorPalette.text.primary,
  secondaryText: ColorPalette.text.secondary,
  titleColor: ColorPalette.text.primary,
  contrast: ColorPalette.white,
  warn: ColorPalette.red,
  disabled: ColorPalette.grey,
  background: ColorPalette.background.bright,
  textOnBackground: ColorPalette.white,
}

export interface ITheme {
  readonly primary: string
  readonly text: string
  readonly secondaryText: string
  readonly titleColor: string
  readonly contrast: string
  readonly warn: string
  readonly disabled: string
  readonly background: string
  readonly textOnBackground: string
  readonly [key: string]: string
}

/**
 * All elements support theming by default, and therefore every element must be wrapped inside a ThemeProvider.
 * The ThemeProvider allows you to define the default colors for most elements.
 *
 * **Example**: If you want all you buttons to be red, instead of writing <Button color="red" /> all the time, you might want to set the "primary" color of your theme to red.
 **/
const ThemeProvider: FunctionComponent<{ theme?: Partial<ITheme> }> = ({
  children,
  theme = defaultTheme,
}) => (
  <ThemeConsumer>
    {contextTheme => (
      <ThemeContext.Provider value={{ ...contextTheme, ...(theme as ITheme) }}>
        {children}
      </ThemeContext.Provider>
    )}
  </ThemeConsumer>
)

export default ThemeProvider
export const ThemeContext = createContext<ITheme>(defaultTheme)
export const ThemeConsumer = ThemeContext.Consumer
