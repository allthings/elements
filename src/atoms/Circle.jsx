import React from 'react'
import PropTypes from 'prop-types'
import View from '../atoms/View'
import { css } from 'glamor'
import Theme from '../behaviour/Theme'

const circle = (backgroundColor, outline, outlineColor, fill, radius) =>
  css({
    border: outline && `1px solid ${outlineColor || backgroundColor}`,
    backgroundColor: fill && backgroundColor,
    borderRadius: '50%',
    height: radius,
    width: radius,
    transition: 'background-color 120ms linear',
  })

/** Hello, I'm a circle */
const Circle = ({
  fill = true,
  color = 'primary',
  outline,
  outlineColor = 'primary',
  radius = 40,
  children,
  ...props
}) => (
  <Theme>
    {({ colorize }) => (
      <View
        {...circle(
          colorize(color),
          outline,
          colorize(outlineColor),
          fill,
          radius
        )}
        alignV="center"
        alignH="center"
        direction="column"
        {...props}
      >
        {children}
      </View>
    )}
  </Theme>
)

Circle.propTypes = {
  /** Content of the Circle will be always centered */
  children: PropTypes.node,
  /** Color of the Circle. Allows theme names (like 'primary') or hex colors  */
  color: PropTypes.string,
  /** Radius of Circle, defaults to 40 */
  radius: PropTypes.number,
  /** If true it will only render the outline */
  outline: PropTypes.bool,
  /** If passed outline will use as outline color instead of color */
  outlineColor: PropTypes.string,
  /** If true it will fill the circle */
  fill: PropTypes.bool,
}

export default Circle
