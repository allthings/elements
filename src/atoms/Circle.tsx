import React from 'react'
import PropTypes from 'prop-types'
import View from './View'
import { css } from 'glamor'
import Theme from '../behaviour/Theme'

const circle = (backgroundColor, outline, outlineColor, fill, radius) =>
  css({
    border: outline && `1px solid ${outlineColor || backgroundColor}`,
    backgroundColor: fill && backgroundColor,
    borderRadius: '50%',
    height: radius,
    width: radius,
    transition: 'background-color 120ms linear',
  })

/**
 * Simple circle element that is used to build other things.
 *
 * Check out ProfileImage to see it in action.
 *
 * ```example
 * <ThemeProvider>
 *   <ResourceProvider>
 *     <View direction="row" alignV="center">
 *       <Circle outline fill={false} outlineColor="lightGrey">
 *         <Icon size="s" name="remove-filled" color="lightGrey" />
 *       </Circle>
 *       <Circle color="primary" />
 *       <Circle color="#bada55" radius={20} />
 *       <Circle color="gray" radius={10} />
 *     </View>
 *   </ResourceProvider>
 * </ThemeProvider>
 * ```
 **/
const Circle = ({
  fill = true,
  color = 'primary',
  outline,
  outlineColor = 'primary',
  radius = 40,
  children,
  ...props
}) => (
  <Theme>
    {({ colorize }) => (
      <View
        {...circle(
          colorize(color),
          outline,
          colorize(outlineColor),
          fill,
          radius
        )}
        alignV="center"
        alignH="center"
        direction="column"
        {...props}
      >
        {children}
      </View>
    )}
  </Theme>
)

Circle.propTypes = {
  /** Content of the Circle will be always centered */
  children: PropTypes.node,
  /** Color of the Circle. Allows theme names (like 'primary') or hex colors  */
  color: PropTypes.string,
  /** Radius of Circle, defaults to 40 */
  radius: PropTypes.number,
  /** If true it will only render the outline */
  outline: PropTypes.bool,
  /** If passed outline will use as outline color instead of color */
  outlineColor: PropTypes.string,
  /** If true it will fill the circle */
  fill: PropTypes.bool,
}

export default Circle
