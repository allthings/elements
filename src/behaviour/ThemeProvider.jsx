import React from 'react'
import PropTypes from 'prop-types'
import { Broadcast, Subscriber } from 'react-broadcast'
import { ColorPalette } from '@allthings/colors'

export const THEME_CHANNEL = 'theme'

export const defaultTheme = {
  primary: ColorPalette.primary.blue,
  text: ColorPalette.text.primary,
  secondaryText: ColorPalette.text.primary,
  titleColor: ColorPalette.text.primary,
  contrast: ColorPalette.white,
  warn: ColorPalette.red,
  disabled: ColorPalette.grey,
  background: ColorPalette.background.bright,
  textOnBackground: ColorPalette.white,
}

/**
* All elements support theming by default, and therefore every element must be wrapped inside a ThemeProvider.
* The ThemeProvider allows you to define the default colors for most elements.
*
* **Example**: If you want all you buttons to be red, instead of writing <Button color="red" /> all the time, you might want to set the "primary" color of your theme to red.
**/
export default class ThemeProvider extends React.PureComponent {
  static propTypes = {
    root: PropTypes.bool,
    theme: PropTypes.shape({
      primary: PropTypes.string,
      text: PropTypes.string,
      secondaryText: PropTypes.string,
      titleColor: PropTypes.string,
      contrast: PropTypes.string,
      warn: PropTypes.string,
      disabled: PropTypes.string,
      background: PropTypes.string,
      textOnBackground: PropTypes.string,
    }),
    children: PropTypes.node,
  }

  static defaultProps = {
    theme: defaultTheme,
  }

  static contextTypes = Subscriber.contextTypes

  render() {
    let root = true
    if (this.context.broadcasts && this.context.broadcasts[THEME_CHANNEL]) {
      root = false
    }
    const { theme, children } = this.props
    return root ? (
      <Broadcast
        channel={THEME_CHANNEL}
        value={{
          ...ThemeProvider.defaultProps.theme,
          ...theme,
        }}
      >
        {children}
      </Broadcast>
    ) : (
      <Subscriber channel={THEME_CHANNEL}>
        {previousTheme => (
          <Broadcast
            channel={THEME_CHANNEL}
            value={{
              ...ThemeProvider.defaultProps.theme,
              ...previousTheme,
              ...theme,
            }}
          >
            {children}
          </Broadcast>
        )}
      </Subscriber>
    )
  }
}

export const ThemeShape = ThemeProvider.propTypes.theme

export const withTheme = (mapThemeToProps, displayName) => WrappedComponent =>
  // eslint-disable-next-line
  class extends React.PureComponent {
    static displayName = displayName || WrappedComponent.displayName
    static component = WrappedComponent

    renderComponent = theme => {
      const props = !mapThemeToProps
        ? { theme }
        : mapThemeToProps(theme, this.props)

      return <WrappedComponent {...this.props} {...props} />
    }

    render() {
      return (
        <Subscriber channel={THEME_CHANNEL}>{this.renderComponent}</Subscriber>
      )
    }
  }
