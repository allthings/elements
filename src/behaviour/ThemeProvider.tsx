import React, { FunctionComponent, createContext } from 'react'
import { ColorPalette } from '@allthings/colors'

export const defaultTheme = {
  primary: ColorPalette.primary.blue,
  text: ColorPalette.text.primary,
  secondaryText: ColorPalette.text.secondary,
  titleColor: ColorPalette.text.primary,
  contrast: ColorPalette.white,
  warn: ColorPalette.red,
  disabled: ColorPalette.grey,
  background: ColorPalette.background.bright,
  textOnBackground: ColorPalette.white,
}

export interface ITheme {
  primary: string
  text: string
  secondaryText: string
  titleColor: string
  contrast: string
  warn: string
  disabled: string
  background: string
  textOnBackground: string
  [key: string]: string
}

const ThemeContext = createContext<ITheme>(defaultTheme)
export const ThemeConsumer = ThemeContext.Consumer

/**
 * All elements support theming by default, and therefore every element must be wrapped inside a ThemeProvider.
 * The ThemeProvider allows you to define the default colors for most elements.
 *
 * **Example**: If you want all you buttons to be red, instead of writing <Button color="red" /> all the time, you might want to set the "primary" color of your theme to red.
 **/
const ThemeProvider: FunctionComponent<{ theme?: Partial<ITheme> }> = ({
  children,
  theme = {},
}) => (
  <ThemeConsumer>
    {contextTheme => (
      <ThemeContext.Provider value={{ ...contextTheme, ...(theme as ITheme) }}>
        {children}
      </ThemeContext.Provider>
    )}
  </ThemeConsumer>
)

export default ThemeProvider

type mapThemeToPropsType = (theme: ITheme, props: any) => any

function createThemeConsumer<T>(
  mapThemeToProps: mapThemeToPropsType,
  WrappedComponent: React.ComponentType<T>,
  props: Omit<T, 'theme'>
) {
  return (theme: ITheme) => {
    const extraProps = !mapThemeToProps
      ? { theme }
      : mapThemeToProps(theme, props)

    return <WrappedComponent {...props} {...extraProps} />
  }
}

export function withTheme(mapThemeToProps: mapThemeToPropsType) {
  return function<T extends { theme: ITheme }>(
    WrappedComponent: React.ComponentType<T>
  ): React.ComponentType<Omit<T, 'theme'>> {
    return function(props: Omit<T, 'theme'>) {
      return (
        <ThemeConsumer>
          {createThemeConsumer<T>(mapThemeToProps, WrappedComponent, props)}
        </ThemeConsumer>
      )
    }
  }
}
