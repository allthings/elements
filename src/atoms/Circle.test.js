import React from 'react'
import renderer from 'react-test-renderer'
import Circle from './Circle'
import ThemeProvider from '../behaviour/ThemeProvider'

test('Circle renders without error', () => {
  const tree = renderer
    .create(
      <ThemeProvider>
        <Circle color="pink" fill={true} radius={5} />
      </ThemeProvider>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
