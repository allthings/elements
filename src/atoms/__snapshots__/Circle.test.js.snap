// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Circle renders without error 1`] = `
.css-1jzdjv2,
[data-css-1jzdjv2] {
  background-color: pink;
  border-radius: 50%;
  height: 5px;
  width: 5px;
  transition: background-color 120ms linear;
  -webkit-transition: background-color 120ms linear;
  -moz-transition: background-color 120ms linear;
}

.css-yirehu,
[data-css-yirehu] {
  box-sizing: border-box;
  align-content: center;
  align-items: center;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: column;
  justify-content: center;
  flex: 0 0 auto;
  -webkit-align-content: center;
  -webkit-box-align: center;
  -webkit-align-items: center;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
  -webkit-flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -webkit-flex: 0 0 auto;
}

<div
  data-css-1jzdjv2=""
  data-css-yirehu=""
/>
`;
